### PR TITLE
Toggable grid

### PR DESCRIPTION
**Pull request by mrniket**

### CODE REVIEW

File: UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs [hunk 1]
```diff
@@ -19,16 +19,24 @@
 
 public class LevelBuilderWindow : EditorWindow
 {
+    public static float SectionTitleHeight = 20;
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs]: Should this not be in `OverlayMenu()` instead? Confusing.

File: UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    private void DisplayGridToggle()
+    {
+        GUILayout.BeginHorizontal();
+
+        EditorGUI.BeginChangeCheck();
+        showGrid = EditorGUILayout.Toggle("Grid", showGrid);
+        if (EditorGUI.EndChangeCheck())
+        {
+            GameObject grid = GameObject.FindWithTag("Grid");
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs]: Could we save this reference for next time?

File: UnityProject/Assets/Editor/Tests/ObstacleGeneratorTests.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+using NUnit.Framework;
+using MapFeatures.Obstacles;
+
+namespace AIMMOUnityTest
+{
+    [TestFixture]
+    internal class ObstacleGeneratorTests
+    {
+        [Test]
+        public void TestAdjustedLocation()
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Editor/Tests/ObstacleGeneratorTests.cs]: There's many more edge cases we need to cover here imo.

File: UnityProject/Assets/Editor/Tests/ScoreGeneratorTest.cs [hunk 4]
```diff
@@ -7,19 +7,19 @@ namespace AIMMOUnityTest
...
     internal class ScoreGeneratorTest
-    {            
+    {
         [Test]
         public void TestGenerateScoreFromPrefab()
         {
             GameObject scoreObject = Resources.Load<GameObject>("Prefabs/Scores/score_bluedisk");
-
-            GameObject generatedScore = ScoreGenerator.GenerateScore(scoreObject);
+            GameObject scoreParent = new GameObject();
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Editor/Tests/ScoreGeneratorTest.cs]: Good call!

File: UnityProject/Assets/Scripts/Generators/ObstacleGenerator.cs [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
+        Vector2 adjustedLocation = GetAdjustedLocation(obstaclePrefab);
         GameObject obstacle = Object.Instantiate(
-            obstaclePrefab, 
-            Vector3.zero, 
+            obstaclePrefab,
+            new Vector3(adjustedLocation.x, 0, adjustedLocation.y),
             Quaternion.identity) as GameObject;
         obstacle.transform.SetParent(GameObject.Find("Obstacles").transform, false);
-
+        Debug.Log(obstacle);
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Scripts/Generators/ObstacleGenerator.cs]: Get rid of this

File: UnityProject/Assets/Scripts/Generators/ObstacleGenerator.cs [hunk 6]
```diff
@@ -20,17 +21,38 @@ public static GameObject GenerateObstacle(ObstacleDTO obstacleDTO)
...
         return obstacle;
     }
+
+    public static Vector2 GetAdjustedLocation(GameObject gameObject)
+    {
+        ObstacleDTO obstacleDTO = new ObstacleDTO();
+        obstacleDTO.location = new Location(x: 0, y: 0);
+        obstacleDTO.width = (int)Mathf.Ceil(gameObject.GetComponent<Renderer>().bounds.size.x);
+        obstacleDTO.height = (int)Mathf.Ceil(gameObject.GetComponent<Renderer>().bounds.size.z);
+        Debug.Log(obstacleDTO.width);
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Scripts/Generators/ObstacleGenerator.cs]: Get rid

File: UnityProject/Assets/Scripts/Generators/OverlayGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    private void UpdateGridMaterial(TerrainDTO terrain)
+    {
+        Material material = gridGameObject.GetComponent<Renderer>().sharedMaterial;
+        material.mainTextureScale = new Vector2(terrain.width, terrain.height);
+    }
+
+    private GameObject GetGridGameObject()
+    {
+        GameObject grid = GameObject.FindWithTag("Grid");
+        if (grid == null)
```
Comment (OlafSzmidt) [inline UnityProject/Assets/Scripts/Generators/OverlayGenerator.cs]: We don't necessarily need to do this, but I would change the `==` to `is` as we can always overwrite the equality operator in the future and then we can get some nasty results here

File: UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs [hunk 8]
```diff
@@ -19,16 +19,24 @@
...
+    public void OnEnable()
     {
-        new TerrainMenu(),
-        new PickupMenu(),
-        new ObstacleMenu(),
-    };
+        menus = new IMenu[]
+        {
+            new TerrainMenu(),
+            new OverlayMenu(),
```
Comment (CelineBoudier) [inline UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs]: bear in mind we'll have different overlays that might be toggable independently, like the compass not only the grid

File: UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor;
+using UnityEngine;
+using MapFeatures.Overlays;
+
+/* 
+ * Allows the user to choose when to display the grid and how big it should be.
```
Comment (CelineBoudier) [inline UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs]: not generic enough - right now it's the grid but the menu is Overlay in general?

File: UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs [hunk 10]
```diff
@@ -5,10 +5,13 @@ public class TerrainMenu : IMenu
...
     private Vector2Int terrainSize;
     private TerrainGenerator terrainGenerator;
+    private OverlayGenerator gridGenerator;
 
-    public TerrainMenu() {
+    public TerrainMenu()
+    {
         terrainSize = new Vector2Int(10, 10);
         terrainGenerator = new TerrainGenerator();
+        gridGenerator = new OverlayGenerator();
```
Comment (CelineBoudier) [inline UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs]: consistency?

File: UnityProject/Assets/Editor/Tests/ObstacleGeneratorTests.cs (hunk 3 above)
Comment (CelineBoudier) [reply UnityProject/Assets/Editor/Tests/ObstacleGeneratorTests.cs]: agree - what does AdjustedLocation mean? It seems that there could be a lot of location related tests?

File: UnityProject/Assets/Editor/Tests/OverlayGeneratorTests.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+            Assert.AreEqual(generatedGrid.transform.transform.localPosition.x, 0);
+            Assert.AreEqual(generatedGrid.transform.transform.localPosition.y, 0.01f);
+            Assert.AreEqual(generatedGrid.transform.transform.localPosition.z, 0);
+            Assert.AreEqual(
+                generatedGrid.GetComponent<Renderer>().sharedMaterial.mainTextureScale,
+                new Vector2(81, 2)
+            );
+        }
+    }
+}
```
Comment (CelineBoudier) [inline UnityProject/Assets/Editor/Tests/OverlayGeneratorTests.cs]: new line

File: UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs (hunk 1 above)
Comment (mrniket) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs]: I need to change the other menus to use this

File: UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs (hunk 8 above)
Comment (mrniket) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/LevelBuilderWindow.cs]: Yeah this is the menu for the section to put all of those toggles

File: UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs (hunk 9 above)
Comment (mrniket) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/OverlayMenu.cs]: correct, will change this

File: UnityProject/Assets/Scripts/Generators/OverlayGenerator.cs (hunk 7 above)
Comment (mrniket) [reply UnityProject/Assets/Scripts/Generators/OverlayGenerator.cs]: C# 7 or higher is needed for this, our project is  in C# 4 so I don't think I can do this

File: UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs (hunk 10 above)
Comment (mrniket) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs]: in this case we are in the terrain menu where I am using the overlaygenerator only as a grid generator (to align it to the terrain). This name provides more information/context but I can see the confusion. What do you think?

Comment (CelineBoudier) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs]: dunno - there's also the CreateGrid() function that creates OverlayGenerator.GenerateGridForTerrain(DEFAULT_TERRAIN); etc

Comment (mrniket) [reply UnityProject/Assets/Editor/LevelBuilder/GUI/Menus/TerrainMenu.cs]: cool, renamed it
